PhysicsTools/PatAlgos/plugins/PATElectronProducer.cc

```diff
@@ -323,6 +336,7 @@ void PATElectronProducer::produce(edm::Event & iEvent, const edm::EventSetup & i
 	  const edm::RefToBase<reco::GsfElectron>& elecsRef = electrons->refAt(idx);
 	  Electron anElectron(elecsRef);
 	  anElectron.setPFCandidateRef( pfRef  );
+    anElectron.setIsolationPUPPI((*PUPPIIsolation_charged_hadrons)[elePtr], (*PUPPIIsolation_neutral_hadrons)[elePtr], (*PUPPIIsolation_photons)[elePtr]);
```

Comment:
The track matching is needed to identify the PF candidate as the electrons in PF are track seeded electrons. And I believe that the option was there for full particle flow object based analysis including electrons and muons. 

The collection of gedGsfElectrons also contains ecal seeded electrons.  Some of these electrons would survive our usual offline cut. So it does not need to be matched to PF candidate unless one really wants to do the analysis with full particle flow objects. Besides, the default value for useParticleFlow is "False".